{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"name":"python","version":"3.11.11","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"},"kaggle":{"accelerator":"none","dataSources":[{"sourceType":"datasetVersion","sourceId":11585316,"datasetId":7264061,"databundleVersionId":12043945},{"sourceType":"modelInstanceVersion","sourceId":11371,"databundleVersionId":7771674,"modelInstanceId":5171}],"dockerImageVersionId":31012,"isInternetEnabled":false,"language":"python","sourceType":"notebook","isGpuEnabled":false}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"code","source":"!pip install -U keras-hub\n!pip install -U keras\n!pip install rouge-score","metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true,"execution":{"iopub.status.busy":"2025-04-27T13:03:51.894918Z","iopub.execute_input":"2025-04-27T13:03:51.895233Z","iopub.status.idle":"2025-04-27T13:14:00.928137Z","shell.execute_reply.started":"2025-04-27T13:03:51.895205Z","shell.execute_reply":"2025-04-27T13:14:00.926935Z"}},"outputs":[{"name":"stdout","text":"Requirement already satisfied: keras-hub in /usr/local/lib/python3.11/dist-packages (0.18.1)\n\u001b[33mWARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7a9db11faf50>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/keras-hub/\u001b[0m\u001b[33m\n\u001b[0m\u001b[33mWARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7a9db11f8550>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/keras-hub/\u001b[0m\u001b[33m\n\u001b[0m\u001b[33mWARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7a9db11dbed0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/keras-hub/\u001b[0m\u001b[33m\n\u001b[0m\u001b[33mWARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7a9db11d8550>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/keras-hub/\u001b[0m\u001b[33m\n\u001b[0m\u001b[33mWARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7a9db11bb250>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/keras-hub/\u001b[0m\u001b[33m\n\u001b[0mRequirement already satisfied: absl-py in /usr/local/lib/python3.11/dist-packages (from keras-hub) (1.4.0)\nRequirement already satisfied: numpy in /usr/local/lib/python3.11/dist-packages (from keras-hub) (1.26.4)\nRequirement already satisfied: packaging in /usr/local/lib/python3.11/dist-packages (from keras-hub) (24.2)\nRequirement already satisfied: regex in /usr/local/lib/python3.11/dist-packages (from keras-hub) (2024.11.6)\nRequirement already satisfied: rich in /usr/local/lib/python3.11/dist-packages (from keras-hub) (14.0.0)\nRequirement already satisfied: kagglehub in /usr/local/lib/python3.11/dist-packages (from keras-hub) (0.3.11)\nRequirement already satisfied: tensorflow-text in /usr/local/lib/python3.11/dist-packages (from keras-hub) (2.18.1)\nRequirement already satisfied: pyyaml in /usr/local/lib/python3.11/dist-packages (from kagglehub->keras-hub) (6.0.2)\nRequirement already satisfied: requests in /usr/local/lib/python3.11/dist-packages (from kagglehub->keras-hub) (2.32.3)\nRequirement already satisfied: tqdm in /usr/local/lib/python3.11/dist-packages (from kagglehub->keras-hub) (4.67.1)\nRequirement already satisfied: mkl_fft in /usr/local/lib/python3.11/dist-packages (from numpy->keras-hub) (1.3.8)\nRequirement already satisfied: mkl_random in /usr/local/lib/python3.11/dist-packages (from numpy->keras-hub) (1.2.4)\nRequirement already satisfied: mkl_umath in /usr/local/lib/python3.11/dist-packages (from numpy->keras-hub) (0.1.1)\nRequirement already satisfied: mkl in /usr/local/lib/python3.11/dist-packages (from numpy->keras-hub) (2025.1.0)\nRequirement already satisfied: tbb4py in /usr/local/lib/python3.11/dist-packages (from numpy->keras-hub) (2022.1.0)\nRequirement already satisfied: mkl-service in /usr/local/lib/python3.11/dist-packages (from numpy->keras-hub) (2.4.1)\nRequirement already satisfied: markdown-it-py>=2.2.0 in /usr/local/lib/python3.11/dist-packages (from rich->keras-hub) (3.0.0)\nRequirement already satisfied: pygments<3.0.0,>=2.13.0 in /usr/local/lib/python3.11/dist-packages (from rich->keras-hub) (2.19.1)\nRequirement already satisfied: tensorflow<2.19,>=2.18.0 in /usr/local/lib/python3.11/dist-packages (from tensorflow-text->keras-hub) (2.18.0)\nRequirement already satisfied: mdurl~=0.1 in /usr/local/lib/python3.11/dist-packages (from markdown-it-py>=2.2.0->rich->keras-hub) (0.1.2)\nRequirement already satisfied: astunparse>=1.6.0 in /usr/local/lib/python3.11/dist-packages (from tensorflow<2.19,>=2.18.0->tensorflow-text->keras-hub) (1.6.3)\nRequirement already satisfied: flatbuffers>=24.3.25 in /usr/local/lib/python3.11/dist-packages (from tensorflow<2.19,>=2.18.0->tensorflow-text->keras-hub) (25.2.10)\nRequirement already satisfied: gast!=0.5.0,!=0.5.1,!=0.5.2,>=0.2.1 in /usr/local/lib/python3.11/dist-packages (from tensorflow<2.19,>=2.18.0->tensorflow-text->keras-hub) (0.6.0)\nRequirement already satisfied: google-pasta>=0.1.1 in /usr/local/lib/python3.11/dist-packages (from tensorflow<2.19,>=2.18.0->tensorflow-text->keras-hub) (0.2.0)\nRequirement already satisfied: libclang>=13.0.0 in /usr/local/lib/python3.11/dist-packages (from tensorflow<2.19,>=2.18.0->tensorflow-text->keras-hub) (18.1.1)\nRequirement already satisfied: opt-einsum>=2.3.2 in /usr/local/lib/python3.11/dist-packages (from tensorflow<2.19,>=2.18.0->tensorflow-text->keras-hub) (3.4.0)\nRequirement already satisfied: protobuf!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<6.0.0dev,>=3.20.3 in /usr/local/lib/python3.11/dist-packages (from tensorflow<2.19,>=2.18.0->tensorflow-text->keras-hub) (3.20.3)\nRequirement already satisfied: setuptools in /usr/local/lib/python3.11/dist-packages (from tensorflow<2.19,>=2.18.0->tensorflow-text->keras-hub) (75.1.0)\nRequirement already satisfied: six>=1.12.0 in /usr/local/lib/python3.11/dist-packages (from tensorflow<2.19,>=2.18.0->tensorflow-text->keras-hub) (1.17.0)\nRequirement already satisfied: termcolor>=1.1.0 in /usr/local/lib/python3.11/dist-packages (from tensorflow<2.19,>=2.18.0->tensorflow-text->keras-hub) (2.5.0)\nRequirement already satisfied: typing-extensions>=3.6.6 in /usr/local/lib/python3.11/dist-packages (from tensorflow<2.19,>=2.18.0->tensorflow-text->keras-hub) (4.13.1)\nRequirement already satisfied: wrapt>=1.11.0 in /usr/local/lib/python3.11/dist-packages (from tensorflow<2.19,>=2.18.0->tensorflow-text->keras-hub) (1.17.2)\nRequirement already satisfied: grpcio<2.0,>=1.24.3 in /usr/local/lib/python3.11/dist-packages (from tensorflow<2.19,>=2.18.0->tensorflow-text->keras-hub) (1.70.0)\nRequirement already satisfied: tensorboard<2.19,>=2.18 in /usr/local/lib/python3.11/dist-packages (from tensorflow<2.19,>=2.18.0->tensorflow-text->keras-hub) (2.18.0)\nRequirement already satisfied: keras>=3.5.0 in /usr/local/lib/python3.11/dist-packages (from tensorflow<2.19,>=2.18.0->tensorflow-text->keras-hub) (3.5.0)\nRequirement already satisfied: h5py>=3.11.0 in /usr/local/lib/python3.11/dist-packages (from tensorflow<2.19,>=2.18.0->tensorflow-text->keras-hub) (3.12.1)\nRequirement already satisfied: ml-dtypes<0.5.0,>=0.4.0 in /usr/local/lib/python3.11/dist-packages (from tensorflow<2.19,>=2.18.0->tensorflow-text->keras-hub) (0.4.1)\nRequirement already satisfied: tensorflow-io-gcs-filesystem>=0.23.1 in /usr/local/lib/python3.11/dist-packages (from tensorflow<2.19,>=2.18.0->tensorflow-text->keras-hub) (0.37.1)\nRequirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.11/dist-packages (from requests->kagglehub->keras-hub) (3.4.1)\nRequirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.11/dist-packages (from requests->kagglehub->keras-hub) (3.10)\nRequirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.11/dist-packages (from requests->kagglehub->keras-hub) (2.3.0)\nRequirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.11/dist-packages (from requests->kagglehub->keras-hub) (2025.1.31)\nRequirement already satisfied: intel-openmp<2026,>=2024 in /usr/local/lib/python3.11/dist-packages (from mkl->numpy->keras-hub) (2024.2.0)\nRequirement already satisfied: tbb==2022.* in /usr/local/lib/python3.11/dist-packages (from mkl->numpy->keras-hub) (2022.1.0)\nRequirement already satisfied: tcmlib==1.* in /usr/local/lib/python3.11/dist-packages (from tbb==2022.*->mkl->numpy->keras-hub) (1.2.0)\nRequirement already satisfied: intel-cmplr-lib-rt in /usr/local/lib/python3.11/dist-packages (from mkl_umath->numpy->keras-hub) (2024.2.0)\nRequirement already satisfied: wheel<1.0,>=0.23.0 in /usr/local/lib/python3.11/dist-packages (from astunparse>=1.6.0->tensorflow<2.19,>=2.18.0->tensorflow-text->keras-hub) (0.45.1)\nRequirement already satisfied: intel-cmplr-lib-ur==2024.2.0 in /usr/local/lib/python3.11/dist-packages (from intel-openmp<2026,>=2024->mkl->numpy->keras-hub) (2024.2.0)\nRequirement already satisfied: namex in /usr/local/lib/python3.11/dist-packages (from keras>=3.5.0->tensorflow<2.19,>=2.18.0->tensorflow-text->keras-hub) (0.0.8)\nRequirement already satisfied: optree in /usr/local/lib/python3.11/dist-packages (from keras>=3.5.0->tensorflow<2.19,>=2.18.0->tensorflow-text->keras-hub) (0.14.0)\nRequirement already satisfied: markdown>=2.6.8 in /usr/local/lib/python3.11/dist-packages (from tensorboard<2.19,>=2.18->tensorflow<2.19,>=2.18.0->tensorflow-text->keras-hub) (3.7)\nRequirement already satisfied: tensorboard-data-server<0.8.0,>=0.7.0 in /usr/local/lib/python3.11/dist-packages (from tensorboard<2.19,>=2.18->tensorflow<2.19,>=2.18.0->tensorflow-text->keras-hub) (0.7.2)\nRequirement already satisfied: werkzeug>=1.0.1 in /usr/local/lib/python3.11/dist-packages (from tensorboard<2.19,>=2.18->tensorflow<2.19,>=2.18.0->tensorflow-text->keras-hub) (3.1.3)\nRequirement already satisfied: MarkupSafe>=2.1.1 in /usr/local/lib/python3.11/dist-packages (from werkzeug>=1.0.1->tensorboard<2.19,>=2.18->tensorflow<2.19,>=2.18.0->tensorflow-text->keras-hub) (3.0.2)\nRequirement already satisfied: keras in /usr/local/lib/python3.11/dist-packages (3.5.0)\n\u001b[33mWARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7d7d27073550>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/keras/\u001b[0m\u001b[33m\n\u001b[0m\u001b[33mWARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7d7d2705cfd0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/keras/\u001b[0m\u001b[33m\n\u001b[0m\u001b[33mWARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7d7d2705e790>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/keras/\u001b[0m\u001b[33m\n\u001b[0m\u001b[33mWARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7d7d27150dd0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/keras/\u001b[0m\u001b[33m\n\u001b[0m\u001b[33mWARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7d7d2703df10>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/keras/\u001b[0m\u001b[33m\n\u001b[0mRequirement already satisfied: absl-py in /usr/local/lib/python3.11/dist-packages (from keras) (1.4.0)\nRequirement already satisfied: numpy in /usr/local/lib/python3.11/dist-packages (from keras) (1.26.4)\nRequirement already satisfied: rich in /usr/local/lib/python3.11/dist-packages (from keras) (14.0.0)\nRequirement already satisfied: namex in /usr/local/lib/python3.11/dist-packages (from keras) (0.0.8)\nRequirement already satisfied: h5py in /usr/local/lib/python3.11/dist-packages (from keras) (3.12.1)\nRequirement already satisfied: optree in /usr/local/lib/python3.11/dist-packages (from keras) (0.14.0)\nRequirement already satisfied: ml-dtypes in /usr/local/lib/python3.11/dist-packages (from keras) (0.4.1)\nRequirement already satisfied: packaging in /usr/local/lib/python3.11/dist-packages (from keras) (24.2)\nRequirement already satisfied: mkl_fft in /usr/local/lib/python3.11/dist-packages (from numpy->keras) (1.3.8)\nRequirement already satisfied: mkl_random in /usr/local/lib/python3.11/dist-packages (from numpy->keras) (1.2.4)\nRequirement already satisfied: mkl_umath in /usr/local/lib/python3.11/dist-packages (from numpy->keras) (0.1.1)\nRequirement already satisfied: mkl in /usr/local/lib/python3.11/dist-packages (from numpy->keras) (2025.1.0)\nRequirement already satisfied: tbb4py in /usr/local/lib/python3.11/dist-packages (from numpy->keras) (2022.1.0)\nRequirement already satisfied: mkl-service in /usr/local/lib/python3.11/dist-packages (from numpy->keras) (2.4.1)\nRequirement already satisfied: typing-extensions>=4.5.0 in /usr/local/lib/python3.11/dist-packages (from optree->keras) (4.13.1)\nRequirement already satisfied: markdown-it-py>=2.2.0 in /usr/local/lib/python3.11/dist-packages (from rich->keras) (3.0.0)\nRequirement already satisfied: pygments<3.0.0,>=2.13.0 in /usr/local/lib/python3.11/dist-packages (from rich->keras) (2.19.1)\nRequirement already satisfied: mdurl~=0.1 in /usr/local/lib/python3.11/dist-packages (from markdown-it-py>=2.2.0->rich->keras) (0.1.2)\nRequirement already satisfied: intel-openmp<2026,>=2024 in /usr/local/lib/python3.11/dist-packages (from mkl->numpy->keras) (2024.2.0)\nRequirement already satisfied: tbb==2022.* in /usr/local/lib/python3.11/dist-packages (from mkl->numpy->keras) (2022.1.0)\nRequirement already satisfied: tcmlib==1.* in /usr/local/lib/python3.11/dist-packages (from tbb==2022.*->mkl->numpy->keras) (1.2.0)\nRequirement already satisfied: intel-cmplr-lib-rt in /usr/local/lib/python3.11/dist-packages (from mkl_umath->numpy->keras) (2024.2.0)\nRequirement already satisfied: intel-cmplr-lib-ur==2024.2.0 in /usr/local/lib/python3.11/dist-packages (from intel-openmp<2026,>=2024->mkl->numpy->keras) (2024.2.0)\n\u001b[33mWARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f1ca649f3d0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/rouge-score/\u001b[0m\u001b[33m\n\u001b[0m\u001b[33mWARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f1ca5db4210>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/rouge-score/\u001b[0m\u001b[33m\n\u001b[0m\u001b[33mWARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f1ca5dcced0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/rouge-score/\u001b[0m\u001b[33m\n\u001b[0m\u001b[33mWARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f1ca5de2d10>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/rouge-score/\u001b[0m\u001b[33m\n\u001b[0m\u001b[33mWARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f1ca5df6ed0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/rouge-score/\u001b[0m\u001b[33m\n\u001b[0m\u001b[31mERROR: Could not find a version that satisfies the requirement rouge-score (from versions: none)\u001b[0m\u001b[31m\n\u001b[0m\u001b[31mERROR: No matching distribution found for rouge-score\u001b[0m\u001b[31m\n\u001b[0m","output_type":"stream"}],"execution_count":1},{"cell_type":"markdown","source":"# Download llm model","metadata":{}},{"cell_type":"code","source":"import keras\nimport keras_hub\nimport numpy as np\n","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-04-27T13:14:00.932733Z","iopub.execute_input":"2025-04-27T13:14:00.933083Z","iopub.status.idle":"2025-04-27T13:14:18.910293Z","shell.execute_reply.started":"2025-04-27T13:14:00.933044Z","shell.execute_reply":"2025-04-27T13:14:18.909296Z"}},"outputs":[{"name":"stderr","text":"2025-04-27 13:14:02.684551: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\nWARNING: All log messages before absl::InitializeLog() is called are written to STDERR\nE0000 00:00:1745759642.910098      31 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\nE0000 00:00:1745759642.974078      31 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n","output_type":"stream"}],"execution_count":2},{"cell_type":"code","source":"gemma_lm = keras_hub.models.GemmaCausalLM.from_preset(\"gemma_2b_en\")","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-04-27T13:14:18.911230Z","iopub.execute_input":"2025-04-27T13:14:18.911837Z","iopub.status.idle":"2025-04-27T13:15:49.156028Z","shell.execute_reply.started":"2025-04-27T13:14:18.911811Z","shell.execute_reply":"2025-04-27T13:15:49.155013Z"}},"outputs":[{"name":"stderr","text":"2025-04-27 13:14:20.242279: E external/local_xla/xla/stream_executor/cuda/cuda_driver.cc:152] failed call to cuInit: INTERNAL: CUDA error: Failed call to cuInit: UNKNOWN ERROR (303)\nnormalizer.cc(51) LOG(INFO) precompiled_charsmap is empty. use identity normalization.\n","output_type":"stream"}],"execution_count":3},{"cell_type":"code","source":"prompt_template = \"\"\"\n<start_of_turn>user\nYou are an intelligent AI specialized in generating SQL queries.\nYour task is to assist users in formulating SQL queries to retrieve specific information from a database.\nPlease provide the SQL query corresponding to the given prompt and context:\n\nPrompt:\nConvert a PostgreSQL query into ClickHouse dialect using PG query\n\nContext:\nPG_QUERY: Select * from metric_store.metrics<end_of_turn>\n<start_of_turn>model\n\"\"\"\ngenerated = gemma_lm.generate(prompt_template, max_length=100)","metadata":{"trusted":true,"execution":{"execution_failed":"2025-04-27T13:03:30.205Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"generated","metadata":{"trusted":true,"execution":{"execution_failed":"2025-04-27T13:03:30.205Z"}},"outputs":[],"execution_count":null},{"cell_type":"markdown","source":"Модель не знает синтаксиса clickhouse, ее будет сбивать промпт, указанный выше. Упростим его:","metadata":{}},{"cell_type":"code","source":"prompt_template = \"\"\"\n<start_of_turn>user\nPlease provide the SQL query corresponding to the given prompt and context:\n\nPrompt:\nConvert a SQL query into other dialect using PG query\n\nContext:\nPG_QUERY: Select * from metric_store.metrics<end_of_turn>\n<start_of_turn>model\n\"\"\"\ngenerated = gemma_lm.generate(prompt_template, max_length=100)","metadata":{"trusted":true,"execution":{"execution_failed":"2025-04-27T13:03:30.205Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"generated","metadata":{"trusted":true,"execution":{"execution_failed":"2025-04-27T13:03:30.205Z"}},"outputs":[],"execution_count":null},{"cell_type":"markdown","source":"Некорректно считать какие-то метрики, зная что модель не умеет решать поставленную задачу. Но нужно замерить score базовой модели, чтобы сравнивать ее с предобученной","metadata":{}},{"cell_type":"markdown","source":"# Prepare dataset","metadata":{}},{"cell_type":"code","source":"from datasets import load_dataset\n\ndataset = load_dataset('csv', data_files='/kaggle/input/sql-to-clickhouse/test_dataset.csv')\ndataset","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-04-27T13:15:49.157087Z","iopub.execute_input":"2025-04-27T13:15:49.157460Z","iopub.status.idle":"2025-04-27T13:16:22.750899Z","shell.execute_reply.started":"2025-04-27T13:15:49.157391Z","shell.execute_reply":"2025-04-27T13:16:22.749975Z"}},"outputs":[{"output_type":"display_data","data":{"text/plain":"Generating train split: 0 examples [00:00, ? examples/s]","application/vnd.jupyter.widget-view+json":{"version_major":2,"version_minor":0,"model_id":"42620dbe54f54ecb8b13a79d9aa65001"}},"metadata":{}},{"execution_count":4,"output_type":"execute_result","data":{"text/plain":"DatasetDict({\n    train: Dataset({\n        features: ['natural_languge', 'pg_db_schema', 'pg_query', 'ch_db_schema', 'ch_query'],\n        num_rows: 531\n    })\n})"},"metadata":{}}],"execution_count":4},{"cell_type":"code","source":"prompt_template = \"\"\"\n<start_of_turn>user\nPlease provide the SQL query corresponding to the given prompt and context:\n\nPrompt:\nConvert a SQL query into other dialect using PG query\n\nContext:\n{sql_context} <end_of_turn>\n<start_of_turn>model\n\"\"\"\n\ndef generate_prompt(data_point):\n    \"\"\"Generate input text based on a prompt, task instruction, (context info), and answer.\n\n    :param data_point: dict: Data point\n    :return: dict: Data point with the added \"prompt\" field\n    \"\"\"\n    sql_context = \"\"\"\n    Natural languge of query: {natural_languge},\n    PostgreSQL query: {pg_query},\n    \"\"\".format(\n        natural_languge=data_point[\"natural_languge\"],\n        pg_query=data_point[\"pg_query\"],\n    )\n    prompt_text = prompt_template.format(sql_context=sql_context)\n    data_point[\"prompt\"] = prompt_text\n\n    return data_point\n\ndataset = dataset.map(generate_prompt)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-04-27T13:16:22.753198Z","iopub.execute_input":"2025-04-27T13:16:22.753798Z","iopub.status.idle":"2025-04-27T13:16:22.831747Z","shell.execute_reply.started":"2025-04-27T13:16:22.753771Z","shell.execute_reply":"2025-04-27T13:16:22.830640Z"}},"outputs":[{"output_type":"display_data","data":{"text/plain":"Map:   0%|          | 0/531 [00:00<?, ? examples/s]","application/vnd.jupyter.widget-view+json":{"version_major":2,"version_minor":0,"model_id":"c5627e4d5e56451987ab079f12d4b0b0"}},"metadata":{}}],"execution_count":5},{"cell_type":"code","source":"test_dataset = dataset['train']","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-04-27T13:16:22.832769Z","iopub.execute_input":"2025-04-27T13:16:22.833093Z","iopub.status.idle":"2025-04-27T13:16:22.837816Z","shell.execute_reply.started":"2025-04-27T13:16:22.833068Z","shell.execute_reply":"2025-04-27T13:16:22.836670Z"}},"outputs":[],"execution_count":6},{"cell_type":"markdown","source":"# Calculate model scores","metadata":{}},{"cell_type":"code","source":"import re\nfrom collections import Counter\nimport numpy as np\n\ndef get_ngrams(text, n):\n    \"\"\"Извлекает n-граммы из текста (слова разделяются по пробелам)\"\"\"\n    words = re.findall(r'\\w+', text.lower())\n    return [' '.join(words[i:i+n]) for i in range(len(words)-n+1)]\n\ndef rouge_n(reference, candidate, n=1):\n    \"\"\"\n    Вычисляет ROUGE-N (Precision, Recall, F1) между эталоном и кандидатом.\n    \"\"\"\n    ref_ngrams = get_ngrams(reference, n)\n    cand_ngrams = get_ngrams(candidate, n)\n    \n    ref_counter = Counter(ref_ngrams)\n    cand_counter = Counter(cand_ngrams)\n    \n    # Общие n-граммы\n    overlap = sum((ref_counter & cand_counter).values())\n    \n    precision = overlap / max(len(cand_ngrams), 1e-10)\n    recall = overlap / max(len(ref_ngrams), 1e-10)\n    f1 = 2 * precision * recall / max(precision + recall, 1e-10)\n    \n    return {'precision': precision, 'recall': recall, 'f1': f1}\n\ndef rouge_l(reference, candidate):\n    \"\"\"\n    Упрощённый ROUGE-L (на основе самой длинной общей подпоследовательности слов)\n    \"\"\"\n    ref_words = re.findall(r'\\w+', reference.lower())\n    cand_words = re.findall(r'\\w+', candidate.lower())\n    \n    # Длина наибольшей общей подпоследовательности (LCS)\n    lcs_length = len(_longest_common_subsequence(ref_words, cand_words))\n    \n    precision = lcs_length / max(len(cand_words), 1e-10)\n    recall = lcs_length / max(len(ref_words), 1e-10)\n    f1 = 2 * precision * recall / max(precision + recall, 1e-10)\n    \n    return {'precision': precision, 'recall': recall, 'f1': f1}\n\ndef _longest_common_subsequence(a, b):\n    \"\"\"Алгоритм LCS для двух списков\"\"\"\n    lengths = [[0] * (len(b)+1) for _ in range(len(a)+1)]\n    for i, x in enumerate(a):\n        for j, y in enumerate(b):\n            if x == y:\n                lengths[i+1][j+1] = lengths[i][j] + 1\n            else:\n                lengths[i+1][j+1] = max(lengths[i+1][j], lengths[i][j+1])\n    \n    result = []\n    i, j = len(a), len(b)\n    while i > 0 and j > 0:\n        if lengths[i][j] == lengths[i-1][j]:\n            i -= 1\n        elif lengths[i][j] == lengths[i][j-1]:\n            j -= 1\n        else:\n            result.append(a[i-1])\n            i -= 1\n            j -= 1\n    return result[::-1]\n\n\ndef evaluate_rouge_manual(dataset, model):\n    results = {'rouge-1': [], 'rouge-2': [], 'rouge-l': []}\n    \n    for item in dataset:\n        generated = model.generate(item[\"prompt\"], max_length=len(item[\"prompt\"]) + 100)\n        model_answer = generated.split(\"model\")[1].strip()\n        reference = item[\"ch_query\"]\n        \n        r1 = rouge_n(reference, generated, 1)\n        r2 = rouge_n(reference, generated, 2)\n        rl = rouge_l(reference, generated)\n        \n        results['rouge-1'].append(r1['f1'])\n        results['rouge-2'].append(r2['f1'])\n        results['rouge-l'].append(rl['f1'])\n\n    avg_results = {\n        metric: np.mean(scores) \n        for metric, scores in results.items()\n    }\n    return avg_results","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-04-27T13:16:22.838937Z","iopub.execute_input":"2025-04-27T13:16:22.839216Z","iopub.status.idle":"2025-04-27T13:16:22.862206Z","shell.execute_reply.started":"2025-04-27T13:16:22.839192Z","shell.execute_reply":"2025-04-27T13:16:22.861233Z"}},"outputs":[],"execution_count":7},{"cell_type":"code","source":"# Применение к вашей модели\nrouge_scores = evaluate_rouge_manual(test_dataset.select(range(40)), gemma_lm)\nprint(\"Средние ROUGE-F1:\")\nfor metric, score in rouge_scores.items():\n    print(f\"{metric}: {score:.4f}\")","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-04-27T13:18:44.434026Z","iopub.execute_input":"2025-04-27T13:18:44.434622Z","iopub.status.idle":"2025-04-27T15:06:59.250737Z","shell.execute_reply.started":"2025-04-27T13:18:44.434595Z","shell.execute_reply":"2025-04-27T15:06:59.249752Z"}},"outputs":[{"name":"stdout","text":"Средние ROUGE-F1:\nrouge-1: 0.1213\nrouge-2: 0.0772\nrouge-l: 0.1157\n","output_type":"stream"}],"execution_count":9},{"cell_type":"code","source":"","metadata":{"trusted":true},"outputs":[],"execution_count":null}]}